// license-webpack-plugin/dist/ModuleProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var LicenseExtractor_1 = require("./LicenseExtractor");
var FileUtils_1 = require("./FileUtils");
var ModuleProcessor = (function () {
    function ModuleProcessor(context, options, errors) {
        var _this = this;
        this.context = context;
        this.options = options;
        this.errors = errors;
        this.modulePrefixes = options.modulesDirectories.map(function (dir) {
            return path.join(_this.context, dir);
        });
        this.licenseExtractor = new LicenseExtractor_1.LicenseExtractor(this.options, this.errors);
    }
    ModuleProcessor.prototype.processFile = function (filename) {
        if (!filename || filename.trim() === '') {
            return null;
        }
        var modulePrefix = this.findModulePrefix(filename);
        if (!this.isFromModuleDirectory(filename, modulePrefix)) {
            return null;
        }
        var packageName = this.extractPackageName(filename, modulePrefix);
        return this.processPackage(packageName, modulePrefix);
    };
    ModuleProcessor.prototype.processPackage = function (packageName, modulePrefix) {
        var isParsed = this.licenseExtractor.parsePackage(packageName, modulePrefix);
        return isParsed ? packageName : null;
    };
    ModuleProcessor.prototype.processExternalPackage = function (packageName) {
        var modulePrefix = null;
        for (var _i = 0, _a = this.modulePrefixes; _i < _a.length; _i++) {
            var currentModulePrefix = _a[_i];
            if (FileUtils_1.FileUtils.isThere(path.join(currentModulePrefix, packageName))) {
                modulePrefix = currentModulePrefix;
                break;
            }
        }
        var isParsed = this.licenseExtractor.parsePackage(packageName, modulePrefix);
        return isParsed ? packageName : null;
    };
    ModuleProcessor.prototype.getPackageInfo = function (packageName) {
        return this.licenseExtractor.getCachedPackage(packageName);
    };
    ModuleProcessor.prototype.extractPackageName = function (filename, modulePrefix) {
        var tokens = filename
            .replace(modulePrefix + path.sep, '')
            .split(path.sep);
        return tokens[0].charAt(0) === '@'
            ? tokens.slice(0, 2).join('/')
            : tokens[0];
    };
    ModuleProcessor.prototype.isFromModuleDirectory = function (filename, modulePrefix) {
        var isPackageFile = false;
        if (modulePrefix) {
            // files like /path/to/node_modules/a.js do not count since they don't belong to any package
            isPackageFile =
                filename.replace(modulePrefix + path.sep, '').indexOf(path.sep) > -1;
        }
        return !!filename && !!modulePrefix && isPackageFile;
    };
    ModuleProcessor.prototype.findModulePrefix = function (filename) {
        for (var _i = 0, _a = this.modulePrefixes; _i < _a.length; _i++) {
            var modulePrefix = _a[_i];
            if (filename.startsWith(modulePrefix)) {
                return modulePrefix;
            }
        }
        return null;
    };
    return ModuleProcessor;
}());
exports.ModuleProcessor = ModuleProcessor;
